src/Common/logger_useful.h

```diff
@@ -83,18 +83,20 @@ namespace
                                                                                                                     \
         std::string_view _format_string;                                                                            \
         std::string _formatted_message;                                                                             \
+        std::vector<std::string> _format_string_args;                                                               \
                                                                                                                     \
         if constexpr (LogTypeInfo::is_static)                                                                       \
         {                                                                                                           \
             formatStringCheckArgsNum(LOG_IMPL_FIRST_ARG(__VA_ARGS__), _nargs - 1);                                  \
             _format_string = ConstexprIfsAreNotIfdefs<LogTypeInfo::is_static>::getStaticFormatString(LOG_IMPL_FIRST_ARG(__VA_ARGS__)); \
+            ConstexprIfsAreNotIfdefs<LogTypeInfo::is_static>::getFormattedArgs(_format_string_args, true, __VA_ARGS__); \
```

Comment:
Looks like it will evaluate `__VA_ARGS__` one more time. See https://github.com/ClickHouse/ClickHouse/issues/52074

Reply:
Yes, I already realized it, thanks for pointing out